crates/bevy_pbr/src/render/gpu_preprocess.rs

```diff
@@ -0,0 +1,305 @@
+        );
+
+        PreprocessPipeline {
+            bind_group_layout,
+            pipeline_id: None,
+        }
+    }
+}
+
+/// A system that specializes the `mesh_preprocess.wgsl` pipeline if necessary.
+pub fn prepare_preprocess_pipeline(
+    pipeline_cache: Res<PipelineCache>,
+    mut pipelines: ResMut<SpecializedComputePipelines<PreprocessPipeline>>,
+    mut preprocess_pipeline: ResMut<PreprocessPipeline>,
+) {
+    if preprocess_pipeline.pipeline_id.is_some() {
+        return;
+    }
+
+    let preprocess_pipeline_id = pipelines.specialize(&pipeline_cache, &preprocess_pipeline, ());
+    preprocess_pipeline.pipeline_id = Some(preprocess_pipeline_id);
+}
+
+/// A system that attaches the mesh uniform buffers to the bind group for the
+/// compute shader.
+pub fn prepare_preprocess_bind_groups(
+    mut commands: Commands,
+    render_device: Res<RenderDevice>,
+    batched_instance_buffers: Res<BatchedInstanceBuffers<MeshUniform, MeshInputUniform>>,
+    pipeline: Res<PreprocessPipeline>,
+) {
+    // Grab the `BatchedInstanceBuffers`.
+    let BatchedInstanceBuffers {
+        data_buffer: ref data_buffer_vec,
+        work_item_buffers: ref index_buffers,
+        current_input_buffer: ref current_input_buffer_vec,
+        previous_input_buffer: ref previous_input_buffer_vec,
+    } = batched_instance_buffers.into_inner();
+
+    let (Some(current_input_buffer), Some(previous_input_buffer), Some(data_buffer)) = (
+        current_input_buffer_vec.buffer(),
+        previous_input_buffer_vec.buffer(),
+        data_buffer_vec.buffer(),
+    ) else {
+        return;
+    };
+
+    for (view, index_buffer_vec) in index_buffers {
+        let Some(index_buffer) = index_buffer_vec.buffer() else {
+            continue;
+        };
+
+        // Don't use `as_entire_binding()` here; the shader reads the array
+        // length and the underlying buffer may be longer than the actual size
+        // of the vector.
+        let index_buffer_size = NonZeroU64::try_from(
+            index_buffer_vec.len() as u64 * u64::from(PreprocessWorkItem::min_size()),
+        )
+        .ok();
+
+        commands
+            .entity(*view)
+            .insert(PreprocessBindGroup(render_device.create_bind_group(
+                "preprocess_bind_group",
+                &pipeline.bind_group_layout,
+                &BindGroupEntries::sequential((
+                    current_input_buffer.as_entire_binding(),
+                    previous_input_buffer.as_entire_binding(),
+                    BindingResource::Buffer(BufferBinding {
+                        buffer: index_buffer,
+                        offset: 0,
+                        size: index_buffer_size,
+                    }),
+                    data_buffer.as_entire_binding(),
+                )),
+            )));
+    }
+}
+
+/// Returns `a / b`, rounded toward positive infinity.
+fn div_round_up(a: usize, b: usize) -> usize {
+    (a + b - 1) / b
+}
```

Comment:
Not needed due to the existence of `div_ceil()`, that I didn't know existed until this review.